whetstone.py

```diff
@@ -32,22 +59,482 @@ def _authorize(self):
+
+    def _write_to_json(self, data):
+        """Outputs the request data to a local json file for inspection."""
+        if os.path.exists(self.filename):
+            os.remove(self.filename)
+        with open(self.filename, "w") as f:
+            json.dump(data, f, indent=2)
+
+    def transform_and_load(self):
+        """Formats raw request data into relational table models and inserts into the db."""
+        data = self.get_all()
+        models = self._preprocess_records(data)
+        for model, records in models.items():
+            if records:
+                df = pd.DataFrame(records)
+                df = df.astype("object")
+                df = self._convert_dates(df)
+                df.rename(columns={"_id": "id"}, inplace=True)
+                self._write_to_db(df, model)
+
+    def _preprocess_records(self, records):
+        """
+        Any preprocessing that needs to be done to records before they are mapped to
+        models. Intended to be overridden by subclasses as needed.
+        """
+        return records
+
+    def _convert_dates(self, df):
+        """Convert date columns to the actual date time."""
+        date_types = {col: "datetime64[ns]" for col in df.columns if col in self.dates}
+        if date_types:
+            df = df.astype(date_types)
+        return df
+
+
+class Users(Whetstone):
+    def __init__(self, sql, qa=False):
+        super().__init__(sql)
+        self.columns = [
+            "_id",
+            "internalId",
+            "activeDistrict",
+            "archivedAt",
+            "created",
+            "email",
+            "inactive",
+            "lastActivity",
+            "lastModified",
+            "locked",
+            "name",
+            "first",
+            "last",
+            "school",
+            "course",
+            "coach",
+        ]
+
+    def _preprocess_records(self, records):
+        models = {"Users": []}
+        for record in records:
+            record["school"] = record.get("defaultInformation").get("school")
+            record["course"] = record.get("defaultInformation").get("course")
+            user = {k: v for (k, v) in record.items() if k in self.columns}
+            models["Users"].append(user)
+        return models
+
+
+class Schools(Whetstone):
+    def __init__(self, sql, qa=False):
+        super().__init__(sql)
+        self.columns = [
+            "_id",
+            "internalId",
+            "name",
+            "abbreviation",
+            "archivedAt",
+            "principal",
+            "gradeSpan",
+            "lowGrade",
+            "highGrade",
+            "district",
+            "phone",
+            "address",
+            "city",
+            "cluster",
+            "region",
+            "state",
+            "zip",
+            "lastModified",
+        ]
+
+    def _preprocess_records(self, records):
+        models = {"Schools": [], "ObservationGroups": [], "ObservationGroupMembers": []}
```

Comment:
I think a slightly better practice here would be to create three lists named after these models, and then create your dictionary at the end. That way you don't risk mistyping the keys when you reference them later. This is less of a risk with the single-model classes, but should probably apply everywhere.

Reply:
@zkagin That seems much more verbose for little gain other than preventing typos? Maybe I'm misunderstanding the intent.